File: F26/Week04/examples/StatusDecoder.cs
using System;

public readonly record struct StatusWord(bool Fault, bool Enabled, byte Mode, byte Level);

public static class StatusDecoder
{
    public static StatusWord Decode(byte value)
    {
        return new StatusWord(
            Fault: (value & 0x80) != 0,
            Enabled: (value & 0x40) != 0,
            Mode: (byte)((value >> 3) & 0x07),
            Level: (byte)(value & 0x07));
    }
}

public static class Program
{
    public static void Main()
    {
        byte[] testValues = new byte[] { 0x00, 0xFF, 0x48, 0xB5 };

        foreach (byte value in testValues)
        {
            StatusWord status = StatusDecoder.Decode(value);
            Console.WriteLine(
                $"0x{value:X2} fault={status.Fault} enabled={status.Enabled} " +
                $"mode={status.Mode} level={status.Level}");
        }
    }
}

File: F26/Week04/examples/status_decoder.py
from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class StatusWord:
    fault: bool
    enabled: bool
    mode: int
    level: int


def decode_status(value: int) -> StatusWord:
    if not 0 <= value <= 0xFF:
        raise ValueError("status byte must be in the range 0..255")

    return StatusWord(
        fault=bool(value & 0x80),
        enabled=bool(value & 0x40),
        mode=(value >> 3) & 0x07,
        level=value & 0x07,
    )


if __name__ == "__main__":
    for value in (0x00, 0xFF, 0x48, 0xB5):
        status = decode_status(value)
        print(
            f"0x{value:02X} fault={status.fault} enabled={status.enabled} "
            f"mode={status.mode} level={status.level}"
        )
